### ISSUE TITLE
55

### ISSUE THREAD

**Issue by SteveHong1901** [SteveHong1901/STAT0023-Coursework-1#9]
import numpy as np
from scipy.stats import t

class MulticlassLogisticRegression:
    ...  # [rest of the code you provided]

    def hypothesis_test(self, X, alpha=0.05):
        """
        Perform a hypothesis test on the logistic regression weights.
        Returns a table of weights, t-statistic, and p-value.
        """
        m, n = X.shape
        k = len(self.classes)  # Number of classes

        # Add bias term to X
        X_bias = np.insert(X, 0, 1, axis=1)
        
        logits = X_bias.dot(self.weights)
        probabilities = self._softmax(logits)
        
        # Compute the Hessian matrix
        R = np.diag((probabilities * (1 - probabilities)).reshape(-1))
        Hessian = X_bias.T.dot(R).dot(X_bias)
        
        # Use Cholesky decomposition to find the inverse of the Hessian matrix
        try:
            chol = np.linalg.cholesky(Hessian)
            cov_matrix = np.linalg.inv(chol.T).dot(np.linalg.inv(chol))
        except np.linalg.LinAlgError:
            print("Matrix is not positive definite, cannot compute covariance using Cholesky decomposition.")
            return
        
        print("Weight\tt-statistic\tp-value")
        for j in range(n + 1):
            for c in range(k):
                std_error = np.sqrt(cov_matrix[j, j])
                t_statistic = self.weights[j, c] / std_error
                df = m - n - 1
                p_value = (1 - t.cdf(abs(t_statistic), df)) * 2  # Two-tailed test
                
                # Print results
                print(f"w_{j,c}\t{t_statistic:.4f}\t\t{p_value:.4f}")
                if p_value < alpha:
                    print(f"* Weight w_{j,c} is statistically significant at alpha = {alpha}")
